pallets/moonbeam-orbiters/src/lib.rs

```diff
@@ -0,0 +1,544 @@
+		OrbiterRewarded {
+			account: T::AccountId,
+			rewards: BalanceOf<T>,
+		},
+		OrbiterRotation {
+			collator: T::AccountId,
+			old_orbiter: Option<T::AccountId>,
+			new_orbiter: Option<T::AccountId>,
+		},
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Add an orbiter in a collator pool
+		#[pallet::weight(T::WeightInfo::collator_add_orbiter())]
+		pub fn collator_add_orbiter(
+			origin: OriginFor<T>,
+			orbiter: <T::Lookup as StaticLookup>::Source,
+		) -> DispatchResult {
+			let collator = ensure_signed(origin)?;
+			let orbiter = T::Lookup::lookup(orbiter)?;
+
+			let mut collator_pool =
+				CollatorsPool::<T>::get(&collator).ok_or(Error::<T>::CollatorNotFound)?;
+			let orbiters = collator_pool.get_orbiters();
+			ensure!(
+				(orbiters.len() as u32) < T::MaxPoolSize::get(),
+				Error::<T>::CollatorPoolTooLarge
+			);
+			if orbiters.iter().any(|orbiter_| orbiter_ == &orbiter) {
+				return Err(Error::<T>::OrbiterAlreadyInPool.into());
+			}
+
+			// Make sure the orbiter has made a deposit. It can be an old orbiter whose deposit
+			// is lower than the current minimum (if the minimum was lower in the past), so we just
+			// have to check that a deposit exists (which means checking that the deposit amount
+			// is not zero).
+			let orbiter_deposit =
+				T::Currency::reserved_balance_named(&T::OrbiterReserveIdentifier::get(), &orbiter);
+			ensure!(
+				orbiter_deposit > BalanceOf::<T>::zero(),
+				Error::<T>::OrbiterDepositNotFound
+			);
+
+			collator_pool.add_orbiter(orbiter.clone());
+			CollatorsPool::<T>::insert(collator.clone(), collator_pool);
+
+			Self::deposit_event(Event::OrbiterJoinCollatorPool {
+				collator: collator,
+				orbiter,
+			});
+
+			Ok(())
+		}
+
+		/// Remove an orbiter from the caller collator pool
+		#[pallet::weight(T::WeightInfo::collator_remove_orbiter())]
+		pub fn collator_remove_orbiter(
+			origin: OriginFor<T>,
+			orbiter: <T::Lookup as StaticLookup>::Source,
+		) -> DispatchResult {
+			let collator = ensure_signed(origin)?;
+			let orbiter = T::Lookup::lookup(orbiter)?;
+
+			let mut collator_pool =
+				CollatorsPool::<T>::get(&collator).ok_or(Error::<T>::CollatorNotFound)?;
+
+			match collator_pool.remove_orbiter(&orbiter) {
+				RemoveOrbiterResult::OrbiterNotFound => {
+					return Err(Error::<T>::OrbiterNotFound.into())
+				}
+				RemoveOrbiterResult::OrbiterRemoved => {
+					Self::deposit_event(Event::OrbiterLeaveCollatorPool {
+						collator: collator.clone(),
+						orbiter,
+					});
+				}
+				RemoveOrbiterResult::OrbiterRemoveScheduled => (),
+			}
+
+			CollatorsPool::<T>::insert(collator, collator_pool);
+			Ok(())
+		}
+
+		/// Remove the caller from the specified collator pool
+		#[pallet::weight(T::WeightInfo::orbiter_leave_collator_pool())]
+		pub fn orbiter_leave_collator_pool(
+			origin: OriginFor<T>,
+			collator: <T::Lookup as StaticLookup>::Source,
+		) -> DispatchResult {
+			let orbiter = ensure_signed(origin)?;
```

Comment:
This code looks almost identical to the previous extrinsic (except the ensure_signed part), maybe we can have a unified internal function for both extrinsics?

Reply:
Done